### PR TITLE
WIP (support for ome.zar)

### PR DESCRIPTION
**Pull request by KateMoreva**
Working over opening the raw.ome.zarr file

### CODE REVIEW

**Comment by constantinpape** [comment]
Thanks for opening the PR. I had a quick look and I can't find anything right now. I will have a closer look tomorrow (and also double check the example data).

One observation: the block copy here: https://github.com/mobie/mobie-viewer-fiji/blob/ede04f8b51c1d45203dfdb4606d6a00de65db540/src/main/java/de/embl/cba/mobie/n5/zarr/N5OMEZarrImageLoader.java#L620
should not be necessary for 3d data. But I don't think that this causes the issues.

@tischi This PR tries to adapt the OMEZarrImageLoader to the new ome-zarr v0.3 spec. See https://github.com/mobie/mobie.github.io/issues/52#issuecomment-856860399 for more details and the example data.
It is compiling and runs, but does not show any data. Can you think of any more changes necessary to make it work for the 3d data?

**Comment by tischi** [comment]
If you provide me with a small protocol how to reproduce this I could work on it.

**Comment by constantinpape** [comment]
I tried to build this locally now, but I get an error that I don't understand:
```
Error:java: java.io.IOException: Unexpected char: '}' from org.scijava.plugin.Plugin from java.io.BufferedInputStream@329bc5e1
  	at org.scijava.annotations.IndexReader.next(IndexReader.java:163)
  	at org.scijava.annotations.AbstractIndexWriter.merge(AbstractIndexWriter.java:144)
  	at org.scijava.annotations.AbstractIndexWriter.write(AbstractIndexWriter.java:96)
  	at org.scijava.annotations.AnnotationProcessor.process(AnnotationProcessor.java:93)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:802)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:713)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1043)
  	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1184)
  	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
  	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
  	at com.sun.tools.javac.main.Main.compile(Main.java:523)
  	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
  	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
  	at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:207)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.compileJava(JavaBuilder.java:486)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.compile(JavaBuilder.java:338)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.doBuild(JavaBuilder.java:263)
  	at org.jetbrains.jps.incremental.java.JavaBuilder.build(JavaBuilder.java:216)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runModuleLevelBuilders(IncProjectBuilder.java:1328)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:1006)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1073)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:967)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:796)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:378)
  	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:178)
  	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:140)
  	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:297)
  	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:130)
  	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler.lambda$channelRead0$0(BuildMain.java:232)
  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  	at java.lang.Thread.run(Thread.java:748)
```

**Comment by constantinpape** [comment]
@tischi
To test this you need to:
- download the test data from https://oc.embl.de/index.php/s/KLTaSgX1c4zgnHP
- Check out this branch and build it (doesn't work for me, see error message above)
- Start fiji and open the file `Covid19-S4-Area2/images/bdv.ome.zarr` (relative to the unpacked test data) with "BigDataViewer>OME ZARR>Open OME ZARR from the file system..."

**Comment by KateMoreva** [comment]
> I tried to build this locally now, but I get an error that I don't understand:
> 
> ```
> Error:java: java.io.IOException: Unexpected char: '}' from org.scijava.plugin.Plugin from java.io.BufferedInputStream@329bc5e1
>   	at org.scijava.annotations.IndexReader.next(IndexReader.java:163)
>   	at org.scijava.annotations.AbstractIndexWriter.merge(AbstractIndexWriter.java:144)
>   	at org.scijava.annotations.AbstractIndexWriter.write(AbstractIndexWriter.java:96)
>   	at org.scijava.annotations.AnnotationProcessor.process(AnnotationProcessor.java:93)
>   	at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:802)
>   	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:713)
>   	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1800(JavacProcessingEnvironment.java:91)
>   	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1043)
>   	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1184)
>   	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
>   	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
>   	at com.sun.tools.javac.main.Main.compile(Main.java:523)
>   	at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
>   	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
>   	at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:207)
>   	at org.jetbrains.jps.incremental.java.JavaBuilder.compileJava(JavaBuilder.java:486)
>   	at org.jetbrains.jps.incremental.java.JavaBuilder.compile(JavaBuilder.java:338)
>   	at org.jetbrains.jps.incremental.java.JavaBuilder.doBuild(JavaBuilder.java:263)
>   	at org.jetbrains.jps.incremental.java.JavaBuilder.build(JavaBuilder.java:216)
>   	at org.jetbrains.jps.incremental.IncProjectBuilder.runModuleLevelBuilders(IncProjectBuilder.java:1328)
>   	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:1006)
>   	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1073)
>   	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:967)
>   	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:796)
>   	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:378)
>   	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:178)
>   	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:140)
>   	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:297)
>   	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:130)
>   	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler.lambda$channelRead0$0(BuildMain.java:232)
>   	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
>   	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
>   	at java.lang.Thread.run(Thread.java:748)
> ```
Could you please tell me a bit more about this bug? When exactly does it occur, just while building?

**Comment by constantinpape** [comment]
> Could you please tell me a bit more about this bug? When exactly does it occur, just while building?

Yes, it occurs during build (with intelliJ on the commit ede04f8b51c1d45203dfdb4606d6a00de65db540). I can't really see which part of the code is causing this.

**Comment by constantinpape** [comment]
@tischi @KateMoreva something is wrong with the test data, I am regenerating it now and will let you know when I have the new version.

**Comment by KateMoreva** [comment]
> @tischi @KateMoreva something is wrong with the test data, I am regenerating it now and will let you know when I have the new version.

Ok, thank you!
As for the bug, I do not know what causes it. 
But it turns out that I may have the wrong version of Imagej-utils. I do not think that the problem is because of the version, but still, I should pull new changes.

**Comment by constantinpape** [comment]
Ok, I know what's going on here now. This is not a problem with the data, which is in fact correct, but with the fact that ome.zarr 0.3 uses a different layout for chunks in zarr now. In a nutshell, zarr now supports nested chunks, see https://github.com/zarr-developers/zarr-python/pull/715 and https://github.com/zarr-developers/zarr-python/pull/716 for details.
(And it turns out even zarr-python is still confused by this, that's why I thought the data was incorrect, see https://github.com/zarr-developers/zarr-python/issues/769#issuecomment-858427474).

This most likely explains why the data is not opened correctly, because we expect the "old" dimension_separator ".". Fortunately this is a parameter in `N5ZarrReader`, see https://github.com/saalfeldlab/n5-zarr/blob/master/src/main/java/org/janelia/saalfeldlab/n5/zarr/N5ZarrReader.java#L98.

We have a local copy of this:
https://github.com/mobie/mobie-viewer-fiji/blob/5019b4f5d3e11517173db1f2bb03c70196eaa17f/src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java#L129
(@tischi why do we duplicate the code here and don't use the upstream saalfeldlab version?)

So I think to fix the issue we only need to find where `N5ZarrReader` is instantiated and call it with `dimensionSeparator="/"`.
(Longer term we should read `dimensionSeparator` from the zarr metadta, where the newer versions of zarr are storing it).

**Comment by constantinpape** [comment]
> As for the bug, I do not know what causes it.
> But it turns out that I may have the wrong version of Imagej-utils. I do not think that the problem is because of the version, but still, I should pull new changes.

Ok, for me this also occured when I pulled the latest master for imagej-utils. Before that it wouldn't build before some other class from imagej-utils was missing. So maybe it is related to that.

**Comment by KateMoreva** [comment]
> Ok, I know what's going on here now. This is not a problem with the data, which is in fact correct, but with the fact that ome.zarr 0.3 uses a different layout for chunks in zarr now. In a nutshell, zarr now supports nested chunks, see [zarr-developers/zarr-python#715](https://github.com/zarr-developers/zarr-python/pull/715) and [zarr-developers/zarr-python#716](https://github.com/zarr-developers/zarr-python/pull/716) for details.
> (And it turns out even zarr-python is still confused by this, that's why I thought the data was incorrect, see [zarr-developers/zarr-python#769 (comment)](https://github.com/zarr-developers/zarr-python/issues/769#issuecomment-858427474)).
> 
> This most likely explains why the data is not opened correctly, because we expect the "old" dimension_separator ".". Fortunately this is a parameter in `N5ZarrReader`, see https://github.com/saalfeldlab/n5-zarr/blob/master/src/main/java/org/janelia/saalfeldlab/n5/zarr/N5ZarrReader.java#L98.
> 
> We have a local copy of this:
> https://github.com/mobie/mobie-viewer-fiji/blob/5019b4f5d3e11517173db1f2bb03c70196eaa17f/src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java#L129
> 
> 
> (@tischi why do we duplicate the code here and don't use the upstream saalfeldlab version?)
> So I think to fix the issue we only need to find where `N5ZarrReader` is instantiated and call it with `dimensionSeparator="/"`.
> (Longer term we should read `dimensionSeparator` from the zarr metadta, where the newer versions of zarr are storing it).

I think I found the place where the `N5ZarrReader` is instantiated (https://github.com/mobie/mobie-viewer-fiji/blob/3f33a6852553ff96f45b3597dce19dd263f6491a/src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java#L167), but so far using  `dimensionSeparator="/"` does not solve the problem. 
I'll look through all the links above and work on the issue.

**Comment by constantinpape** [comment]
Ok, I think a good approach to debugging is to check the actual return values here
https://github.com/mobie/mobie-viewer-fiji/blob/3f33a6852553ff96f45b3597dce19dd263f6491a/src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java#L518
and see if they correspond to the correct chunk file paths, i.e. if these file paths actually exist.

**Comment by tischi** [comment]
> why do we duplicate the code here  (N5ZarrReader) and don't use the upstream saalfeldlab version?

Frankly: no idea! 

I could imagine that maybe I just copied it over to make initial development and deployment easier. If there is no diff with the upstream version then we should use that one.

**Comment by constantinpape** [comment]
> Frankly: no idea!
> 
> I could imagine that maybe I just copied it over to make initial development and deployment easier. If there is no diff with the upstream version then we should use that one.

Ok, let's first figure out what's going on currently and then we can work on cleaning this up a bit by adding the upstream saalfeldlab repo as dependency and removing the duplicated code.

File: src/main/java/de/embl/cba/mobie/n5/zarr/N5OMEZarrImageLoader.java [hunk 1]
```diff
@@ -244,6 +244,7 @@ private Multiscale getMultiscale( String pathName ) throws IOException
 			if ( n5 instanceof N5S3ZarrReader )
 			{
 				final N5S3ZarrReader s3ZarrReader = ( N5S3ZarrReader ) n5;
+				s3ZarrReader.setDimensionSeparator("/");
```
Comment (constantinpape) [inline src/main/java/de/embl/cba/mobie/n5/zarr/N5OMEZarrImageLoader.java:247]: I don't really think this change is necessary, as all the changes here are only done to generate the message for the exception thrown below.

File: src/main/java/de/embl/cba/mobie/n5/zarr/N5S3ZarrReader.java [hunk 2]
```diff
@@ -58,10 +58,15 @@
...
+
+	protected String dimensionSeparator;
 	private final String serviceEndpoint;
 
-	public N5S3ZarrReader( AmazonS3 s3, String serviceEndpoint, String bucketName, String containerPath ) throws IOException
+	public void setDimensionSeparator(String dimensionSeparator) {
+		this.dimensionSeparator = dimensionSeparator;
+	}
+
+	public N5S3ZarrReader(AmazonS3 s3, String serviceEndpoint, String bucketName, String containerPath ) throws IOException
```
Comment (constantinpape) [inline src/main/java/de/embl/cba/mobie/n5/zarr/N5S3ZarrReader.java:69]: I think we should add another constructor which has the dimension separator as argument, similar to this https://github.com/mobie/mobie-viewer-fiji/blob/9b1693c437df77b948ed4ed2e588bafc3d15e3e0/src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java#L92.

File: src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java [hunk 3]
```diff
@@ -164,7 +164,7 @@ public N5ZarrReader( final String basePath, final boolean mapN5DatasetAttributes
 	public N5ZarrReader( final String basePath, final GsonBuilder gsonBuilder) throws IOException
 	{
 
-		this(basePath, gsonBuilder, ".");
+		this(basePath, gsonBuilder, "/");
```
Comment (constantinpape) [inline src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java:167]: Ok, I think this is the change that makes it work.
I think the proper way to fix this is to instead use the constructor that takes the dimensionSeparator as an argument
https://github.com/mobie/mobie-viewer-fiji/blob/5019b4f5d3e11517173db1f2bb03c70196eaa17f/src/main/java/de/embl/cba/mobie/n5/zarr/N5ZarrReader.java#L109
when it is called here
https://github.com/mobie/mobie-viewer-fiji/blob/5019b4f5d3e11517173db1f2bb03c70196eaa17f/src/main/java/de/embl/cba/mobie/n5/zarr/OMEZarrReader.java#L38

**Review by constantinpape — COMMENTED** [review]
Good that you figured out how to make it work!
I think we should revert the changes in `N5ZarrReader` though and instead call the proper constructor to set the dimension separator.

**Comment by constantinpape** [comment]
I also tried to build locally again, but I still get the same error as before, see https://github.com/mobie/mobie-viewer-fiji/pull/297#issuecomment-858401435.

@KateMoreva which imagej-utils commit are you using? I am on the latest one from master (21f5f8bb057f1c10770e21a46d94ea1d09290686)

@tischi Could you maybe try if this PR builds for you locally?

**Comment by KateMoreva** [comment]
> Good that you figured out how to make it work!
> I think we should revert the changes in `N5ZarrReader` though and instead call the proper constructor to set the dimension separator.

> I also tried to build locally again, but I still get the same error as before, see [#297 (comment)](https://github.com/mobie/mobie-viewer-fiji/pull/297#issuecomment-858401435).
> 
> @KateMoreva which imagej-utils commit are you using? I am on the latest one from master (21f5f8bb057f1c10770e21a46d94ea1d09290686)
> 
> @tischi Could you maybe try if this PR builds for you locally?

I am using the same [commit](https://github.com/tischi/imagej-utils/commit/21f5f8bb057f1c10770e21a46d94ea1d09290686)   (21f5f8bb057f1c10770e21a46d94ea1d09290686)

**Comment by KateMoreva** [comment]
A small note on the progress. We can read new ome.zarr both from the file system and s3. But only the new one. I am working on getting the dimensionSeparator from the metadata.

**Comment by constantinpape** [comment]
@KateMoreva I think this PR is superseded by #347. Can we close it?

**Pull request closed without merging**